### ISSUE TITLE
403 error

### ISSUE THREAD

**Issue by PlutusCoder** [serve-and-volley/atp-world-tour-tennis-data#36]
Hi :)

First of all, great job on this project. Very nice work and thanks for making it available. :)

I have been trying to get the "tournaments.py" script to work.  When I try to run it for 2016 - 2017, it returns 0 results for both years.

When I debug it, I noticed the server is returning a "403 forbidden" code when I try to GET the URL (

It looks like a cloudfare bot check or a captcha of some sort.

I've put the server response in the attached file.
[html-error-response.txt](https://github.com/serve-and-volley/atp-world-tour-tennis-data/files/11773191/html-error-response.txt)

I noticed in the code that none of the headers are set or changed before sending the GET url request. By default the headers of the python GET request are as follows:

"headers": {
    "Accept": "*/*", 
    "Accept-Encoding": "gzip, deflate", 
    "Host": "www.atptour.com",
    **"User-Agent": "python-requests/2.31.0"**  
}

The user-agent being "python" seems to advertise ourselves as a bot and so I thought that was the problem.

I have tried setting these custom headers before making the GET request:

HEADERS = {
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7", 
    "Accept-Encoding": "gzip, deflate, br", 
    "Accept-Language": "en-GB,en;q=0.9,fr;q=0.8", 
    "Sec-Ch-Ua": "\"Not.A/Brand\";v=\"8\", \"Chromium\";v=\"114\", \"Google Chrome\";v=\"114\"", 
    "Sec-Ch-Ua-Mobile": "?0", 
    "Sec-Ch-Ua-Platform": "\"Windows\"", 
    "Sec-Fetch-Dest": "document", 
    "Sec-Fetch-Mode": "navigate", 
    "Sec-Fetch-Site": "none", 
    "Sec-Fetch-User": "?1", 
    "Upgrade-Insecure-Requests": "1", 
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36", 
    "Host": "www.atptour.com"
}

page = requests.get(url, headers=HEADERS)

.... but that doesn't work either. I am still getting the 403 error.  It doesn't seem to be an IP issue either. If I try to go to the URL with my browser, it displays fine.

Has something changed on their side? Have you tested this recently? 

Any ideas?

Thanks. :)

**Comment by PlutusCoder** [comment]
Hi again

I've done some more digging and found it is indeed the site that is using cloudflare and captchas etc to detect scrapers.

I've found a fix for this by using [Cloudscraper](https://pypi.org/project/cloudscraper/) to get around their checks. I've got it working on the "tournaments.py" script. I'll have a look at the others over the next few days.

If you'd like me to do a "pull request", let me know. :)
